core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -0,0 +1,705 @@
+
+            return new ArbitraryDistributionTaskSource(
+                    getSourceFragmentToRemoteSourceNodeIdMap(fragment.getRemoteSourceNodes()),
+                    sourceExchanges,
+                    exchangeSourceHandles,
+                    targetPartitionSize);
+        }
+
+        public ArbitraryDistributionTaskSource(
+                Map<PlanFragmentId, PlanNodeId> sourceFragmentToRemoteSourceNodeIdMap,
+                Map<PlanFragmentId, Exchange> sourceExchanges,
+                Multimap<PlanFragmentId, ExchangeSourceHandle> exchangeSourceHandles,
+                DataSize targetPartitionSize)
+        {
+            this.sourceFragmentToRemoteSourceNodeIdMap = ImmutableMap.copyOf(requireNonNull(sourceFragmentToRemoteSourceNodeIdMap, "sourceFragmentToRemoteSourceNodeIdMap is null"));
+            this.sourceExchanges = ImmutableMap.copyOf(requireNonNull(sourceExchanges, "sourceExchanges is null"));
+            this.exchangeSourceHandles = ImmutableListMultimap.copyOf(requireNonNull(exchangeSourceHandles, "exchangeSourceHandles is null"));
+            this.targetPartitionSizeInBytes = requireNonNull(targetPartitionSize, "targetPartitionSize is null").toBytes();
+        }
+
+        @Override
+        public List<TaskDescriptor> getMoreTasks()
+        {
+            NodeRequirements nodeRequirements = new NodeRequirements(Optional.empty(), Optional.empty());
+
+            ImmutableList.Builder<TaskDescriptor> result = ImmutableList.builder();
+            int currentPartitionId = 0;
+
+            ImmutableListMultimap.Builder<PlanNodeId, ExchangeSourceHandle> assignedExchangeSourceHandles = ImmutableListMultimap.builder();
+            long assignedExchangeDataSize = 0;
+
+            for (Map.Entry<PlanFragmentId, ExchangeSourceHandle> entry : exchangeSourceHandles.entries()) {
+                PlanFragmentId sourceFragmentId = entry.getKey();
+                PlanNodeId remoteSourcePlanNodeId = sourceFragmentToRemoteSourceNodeIdMap.get(sourceFragmentId);
+                ExchangeSourceHandle originalExchangeSourceHandle = entry.getValue();
+                Exchange sourceExchange = sourceExchanges.get(sourceFragmentId);
+
+                ExchangeSourceSplitter splitter = sourceExchange.split(originalExchangeSourceHandle, targetPartitionSizeInBytes);
+                ImmutableList.Builder<ExchangeSourceHandle> sourceHandles = ImmutableList.builder();
+                while (true) {
+                    checkState(splitter.isBlocked().isDone(), "not supported");
+                    Optional<ExchangeSourceHandle> next = splitter.getNext();
+                    if (next.isEmpty()) {
+                        break;
+                    }
+                    sourceHandles.add(next.get());
+                }
+
+                for (ExchangeSourceHandle handle : sourceHandles.build()) {
```

Comment:
If the first handle's `sizeInBytes` is larger than `targetPartitionSizeInBytes` (which is possible despite the split, as we don't split the files and size of a single file can exceed `targetPartitionSizeInBytes`), then the first task will have empty `exchangeSourceHandles` to read from